### PR TITLE
[FLINK-32815] Implement internal HASHCODE function

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change

Took over https://github.com/apache/flink/pull/23118 to cover the change with narrower scoped tests.

## Verifying this change

Added tests in HashcodeITCase

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/HashcodeITCase.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Tests for verifying runtime behaviour of {@link BuiltInFunctionDefinitions#HASHCODE}.
+ */
+public class HashcodeITCase {
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/HashcodeITCase.java:65]: Can we use `BuiltInFunctionTestBase`? I feel this class has a lot of boilerplate could that one can simply achieve with a `SELECT HASHCODE(MAP['a', 1, 'b', 2])` as well?

**Review by twalthr — CHANGES_REQUESTED** [review]
Thanks for finishing those function PRs. I would like to simplify the tests though.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/HashcodeITCase.java (hunk 1 above)
Comment (dawidwys) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/HashcodeITCase.java:65]: The reason why I went with a custom solution was to be sure to use `Binary` data. Using `BuiltinFunctionTestBase` we'd use `GenericRowData` and alike.

Comment (twalthr) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/HashcodeITCase.java:65]: Makes sense. Thanks for the explanation.

**Review by twalthr — APPROVED** [review]

**Comment by dawidwys** [comment]
CI seems broken by https://github.com/apache/flink/pull/24022#issuecomment-1888811191

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java [hunk 2]
```diff
@@ -2292,6 +2292,16 @@ ANY, and(logical(LogicalTypeRoot.BOOLEAN), LITERAL)
                     .runtimeProvided()
                     .build();
 
+    public static final BuiltInFunctionDefinition HASHCODE =
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:2295]: sorry for the late comment but I just noticed that this field should be called `INTERNAL_HASHCODE` and be moved down to the other `INTERNAL_` functions

**Pull request merged**